### PR TITLE
Store highest instance instead of highest decided

### PR DESCRIPTION
**Pull request by nkryuchkov**

### CODE REVIEW

File: ibft/storage/store.go [hunk 1]
```diff
@@ -194,6 +195,47 @@ func (i *ibftStorage) GetHighestInstance(identifier []byte) (*specqbft.State, er
 	return ret, nil
 }
 
+// SaveInstance saves the state for the instance
```
Comment (nivBlox) [inline ibft/storage/store.go:198]: please add unit test to check those 2 new func's

File: protocol/v2/qbft/instance/commit.go [hunk 2]
```diff
@@ -45,7 +45,7 @@ func (i *Instance) UponCommit(signedCommit *specqbft.SignedMessage, commitMsgCon
 			return false, nil, nil, errors.Wrap(err, "could not get msg commit data")
 		}
 
-		agg, err := aggregateCommitMsgs(commitMsgs)
+		agg, err := AggregateCommitMsgs(commitMsgs)
```
Comment (nivBlox) [inline protocol/v2/qbft/instance/commit.go:48]: @lior-blox this is a diff from spec FYI.

File: protocol/v2/sync/handlers/decided_history.go [hunk 3]
```diff
@@ -38,7 +40,18 @@ func HistoryHandler(plogger *zap.Logger, storeMap *storage.QBFTStores, reporting
 			if store == nil {
 				return nil, errors.New(fmt.Sprintf("not storage found for type %s", msgID.GetRoleType().String()))
 			}
-			results, err := store.GetDecided(msgID[:], sm.Params.Height[0], sm.Params.Height[1])
+			states, err := store.GetInstance(msgID[:], sm.Params.Height[0], sm.Params.Height[1])
+			results := make([]*specqbft.SignedMessage, 0, len(states))
+			for _, state := range states {
+				_, msgs := state.CommitContainer.LongestUniqueSignersForRoundAndValue(state.Round, state.DecidedValue)
```
Comment (nivBlox) [inline protocol/v2/sync/handlers/decided_history.go:46]: Please use - 
LastPreparedRound instead of just round

Comment (nkryuchkov) [reply protocol/v2/sync/handlers/decided_history.go:46]: done

File: protocol/v2/sync/handlers/last_decided.go [hunk 4]
```diff
@@ -34,7 +35,12 @@ func LastDecidedHandler(plogger *zap.Logger, storeMap *storage.QBFTStores, repor
 			if store == nil {
 				return nil, errors.New(fmt.Sprintf("not storage found for type %s", msgID.GetRoleType().String()))
 			}
-			res, err := store.GetHighestDecided(msgID[:])
+			state, err := store.GetHighestInstance(msgID[:])
+			_, msgs := state.CommitContainer.LongestUniqueSignersForRoundAndValue(state.Round, state.DecidedValue)
```
Comment (nivBlox) [inline protocol/v2/sync/handlers/last_decided.go]: maybe it worth to create 1 func for this and reuse it in those 2 places

File: protocol/v2/qbft/instance/commit.go (hunk 2 above)
Comment (nkryuchkov) [reply protocol/v2/qbft/instance/commit.go:49]: added a note

**Review by nivBlox — APPROVED** [review]

File: protocol/v2/qbft/instance/commit.go (hunk 2 above)
Comment (lior-blox) [reply protocol/v2/qbft/instance/commit.go:49]: @alonmuroch Maybe it should be moved to signedMessage in spec, as its not really an commit function and more of signedMessage function?

File: protocol/v2/qbft/instance/commit.go [hunk 5]
```diff
@@ -60,7 +61,8 @@ func commitQuorumForRoundValue(state *specqbft.State, commitMsgContainer *specqb
 	return state.Share.HasQuorum(len(signers)), msgs, nil
 }
 
-func aggregateCommitMsgs(msgs []*specqbft.SignedMessage) (*specqbft.SignedMessage, error) {
+// AggregateCommitMsgs aggregates commit messages.
+func AggregateCommitMsgs(msgs []*specqbft.SignedMessage) (*specqbft.SignedMessage, error) {
```
Comment (lior-blox) [inline protocol/v2/qbft/instance/commit.go:65]: @alonmuroch Maybe it should be moved to signedMessage in spec, as its not really an commit function and more of signedMessage function?

**Review by nivBlox — APPROVED** [review]

File: protocol/v2/qbft/storage/ibft_store.go [hunk 6]
```diff
@@ -1,29 +1,40 @@
...
 // InstanceStore manages instance data
 type InstanceStore interface {
 	// SaveHighestInstance saves the state for the highest instance
-	SaveHighestInstance(state *specqbft.State) error
+	SaveHighestInstance(instance *StoredInstance) error
 	// GetHighestInstance returns the state for the highest instance
-	GetHighestInstance(identifier []byte) (*specqbft.State, error)
+	GetHighestInstance(identifier []byte) (*StoredInstance, error)
+	// GetInstance returns historical decided messages in the given range
+	GetInstance(identifier []byte, from specqbft.Height, to specqbft.Height) ([]*StoredInstance, error)
```
Comment (amir-blox) [inline protocol/v2/qbft/storage/ibft_store.go:33]: rename to `GetInstances` or `GetInstanceInRange`

Comment (nkryuchkov) [reply protocol/v2/qbft/storage/ibft_store.go]: Is `GetInstanceInRange` ok? Renamed

**Review by lior-blox — APPROVED** [review]

**Pull request merged**